zigzag-circuit.md

```diff
@@ -0,0 +1,122 @@
+# ZigZag: Offline PoRep Circuit Spec
+
+ZigZag is the Proof of Replication used in Filecoin. The prover encodes the original data into a replica and commits to it. An offline PoRep proves that the commitment to the replica is a valid commitment of the encoded original data.
+
+#### ZigZag Overview
+
+ZigZag is the Proof of Replication used in Filecoin. The prover encodes the original data into a replica and commits to it. An offline PoRep proves that the commitment to the replica is a valid commitment of the encoded original data.
+
+ZigZag PoRep is based on layering DRG graphs `l` times. The data represented in each DRG layer is the data encoded in the previous layer. The final layer is the replica (which in Filecoin terms is the sealed sector).
+
+- `replicaId` is a unique replica identifier (see the Filecoin Proofs spec for details)
+- `CommD` is the Merkle Tree root hash of the input data to the first layer
+- `CommR_{i}` is the Merkle Tree hash of the output of the DRG encoding at each layer `i` 
+- `CommRStar` is the hash of the concatenation of the `replicaId` and all the `commR`s.
+
+In the PoRep offline proof as presented in Fisch2019, the proof is large for blockchain usage. We use SNARKs to generate a proof of knowledge of an offline proof. In other words, we use encode the ZigZag proof verification algorithm in an arithmetic circuit and use SNARKs to prove that it was evaluated correctly.
+
+This circuit proves that given a Merkle root `CommD`, `CommR_l`, and `commRStar`, that the prover knew the correct replicated data at each layer.
+
+## Offline PoRep circuit
+
+**Public Parameters**: *Parameters that are embeded in the circuits or used to generate the circuit*
+
+- `LAYERS : UInt`: Number of layers
+- `LAYER_CHALLENGES[0..LAYERS]`: Number of challenges per layer
+- `EXPANSION_DEGREE: UInt`: Degree of the bipartite expander graph to extend dependencies between layers
+- `TREE_DEPTH: UInt`: Height of the Merkle tree
```

Comment:
nit: I'm assuming that *height* and *depth* are used interchangeably here but it would be nice to unify the terminology

Reply:
fixing!